calculator_test.go

```diff
@@ -24,31 +24,122 @@ func TestSubtract(t *testing.T) {
 }
 
 type TestCase struct {
-	A    float64
-	B    float64
+	name string
+	a, b float64
+	f    func(float64, float64) float64
 	want float64
 }
 
-func TestMultiply(t *testing.T) {
+func TestCalculator(t *testing.T) {
 	t.Parallel()
 	testCases := []TestCase{
 		{
-			A:    2,
-			B:    2,
+			name: "Add two positive integers",
+			a:    2, b: 2,
+			f:    calculator.Add,
 			want: 4,
 		},
 		{
-			A:    -3,
-			B:    3,
-			want: -9,
+			name: "Add two negative integers",
+			a:    -2, b: -2,
+			f:    calculator.Add,
+			want: -4,
+		},
+		{
+			name: "Add a positive integer and a negative fraction ",
+			a:    5, b: -2.48,
+			f:    calculator.Add,
+			want: 2.52,
+		},
+		{
+			name: "Subtract a larger integer from a smaller integer",
+			a:    3, b: 5,
+			f:    calculator.Subtract,
+			want: -2,
+		},
+		{
+			name: "Subtract a negative integer from a negative integer",
+			a:    -9, b: -5,
+			f:    calculator.Subtract,
+			want: -4,
+		},
+		{
+			name: "Multiply two positive integers",
+			a:    4, b: 20,
+			f:    calculator.Multiply,
+			want: 80,
+		},
+		{
+			name: "Multiply a positive fraction by a negative fraction",
+			a:    4, b: -9,
+			f:    calculator.Multiply,
+			want: -36,
 		},
 	}
 
 	for _, tc := range testCases {
-		got := calculator.Multiply(tc.A, tc.B)
-		if tc.want != got {
-			t.Errorf("Multiply(%f, %f): want %f, got %f", 2.0, 2.0, tc.want, got)
-		}
+		t.Run(tc.name, func(t *testing.T) {
+			got := tc.f(tc.a, tc.b)
+			if tc.want != got {
+				t.Errorf("want %f, got %f", tc.want, got)
+			}
+		})
+	}
+}
+
+type TestCaseDivide struct {
+	a, b        float64
+	f           func(float64, float64) (float64, error)
+	want        float64
+	errExpected bool
+}
+
+func TestDivide(t *testing.T) {
+	testCases := []TestCaseDivide{
+		{
+			a: 5, b: 0,
+			f:           calculator.Divide,
+			want:        0,
+			errExpected: true,
+		},
+		{
+			a: 10, b: 2.5,
+			f:           calculator.Divide,
+			want:        4,
+			errExpected: false,
+		},
+		{
+			a: 28, b: 2,
+			f:           calculator.Divide,
+			want:        14,
+			errExpected: false,
+		},
+		{
+			a: 1, b: 1,
+			f:           calculator.Divide,
+			want:        1,
+			errExpected: false,
+		},
+		{
+			a: 0, b: 2.5,
+			f:           calculator.Divide,
+			want:        0,
+			errExpected: false,
+		},
 	}
 
+	for _, tc := range testCases {
+		var errorReturned bool = false
+		got, err := tc.f(tc.a, tc.b)
+		if err != nil {
+			errorReturned = true
+		}
+		if tc.errExpected != errorReturned {
+			t.Errorf("Divide(%f, %f): Expected error but got %v", tc.a, tc.b, err)
+		}
+		if got != tc.want {
```

Comment:
And we only want to do this comparison for _valid_ test cases, so we could write:

```go
if !tc.errExpected && tc.want != got {
```